accounts-db/src/accounts_db.rs

```diff
@@ -3674,18 +3674,24 @@ impl AccountsDb {
             "if filtering for incremental snapshots, then snapshots should be enabled",
         );
 
-        purges_zero_lamports.retain(|pubkey, (slot_account_infos, _ref_count)| {
-            // Only keep purges_zero_lamports where the entire history of the account in the root set
-            // can be purged. All AppendVecs for those updates are dead.
-            for (slot, _account_info) in slot_account_infos.iter() {
-                if let Some(store_count) = store_counts.get(slot) {
-                    if store_count.0 != 0 {
-                        // one store this pubkey is in is not being removed, so this pubkey cannot be removed at all
+        purges_zero_lamports.retain(|pubkey, (slot_account_infos, ref_count)| {
+            // a zero lamport account with a single ref count and single slot list entry can be removed completely, even if the slot it is in is not being deleted
+            if !(slot_account_infos.len() == 1
```

Comment:
on line 3714:
```
let cannot_purge = *slot > last_full_snapshot_slot.unwrap();
```
Could we relax the `cannot_purge` by allowing purge this case? 
```
let the_only_zero_lamport = !(slot_account_infos.len() == 1 
                && slot_account_infos.first().unwrap().1.is_zero_lamport()
                && ref_count == &1);
let cannot_purge = *slot > last_full_snapshot_slot.unwrap() && !the_only_zero_lamport;
```

Reply:
Never mind. We can't because "the_only_zero_lamport" is our current view. We can't sure that know when last full snapshot is taken there is not this pubkey.

Reply:
i don't understand. we have to have the top first change (3679 above) or else we bail out early. the second change is also required since we must keep the zero lamport account around if we haven't done a full snapshot recently enough.

Reply:
Yeah. My initial thought is that if we know this is the only zero lamport account, even if slot > last_full_slot, maybe we can clean this account? 
Then, I realize that the only zero lamport account check is our current view of the index. There might be a version of this account stored, when the full snapshot is taken. And we cleaned it after taking the full snapshot. Now our view shows that there is only one zero lamport account. Our current view of the only zero lamport account doesn't guarantee that there is no such account in the full snapshot.